2282307 attachment endpoint

jeqq opened the pull request:
Implementation of Attachment endpoint. 

What is not yet implemented/blockers: 
- For get request the value of "Content-Type" header is not correct in the response headers, it should be the mime type of the attachment.
- On put request the file should be attached to the entity as a new value for field_test_file.

jeqq committed:
Implementation of Attachment endpoint. Initial work.

jeqq committed:
Implement field item normalization.

jeqq committed:
Start to implement head and get methods, add tests for these methods and normalizer.

jeqq committed:
Add _attachments key to normalized content entity, fix FileItemNormalizerTest.

jeqq committed:
Head and Get methods implementation, add more test cases.

jeqq committed:
Replace name with field_name for field_storage_config entity type.

jeqq committed:
Add response for the files that does not live on disk.

jeqq committed:
Add put and delete methods, work in progress.

jeqq committed:
Normalizer and resource fixes.

jeqq committed:
Uncomment commented test cases.

dickolsson commented on relaxed.services.yml:
```diff
@@ -30,6 +30,10 @@ services:
     class: Drupal\relaxed\Normalizer\RevsDiffNormalizer
     tags:
       - { name: normalizer, priority: 20 }
+  relaxed.normalizer.attachment_file_item:
```
Let's just call this service `relaxed.normalizer.file_item`

dickolsson commented on src/Normalizer/ContentEntityNormalizer.php:
```diff
@@ -40,7 +40,22 @@ public function normalize($entity, $format = NULL, array $context = array()) {
+        }
+        continue;
+      }
       $field_data = $this->serializer->normalize($field, $format, $context);
```
It doesn't look nice to call `$field_data = $this->serializer->normalize($field, $format, $context);` twice in `ContentEntitySerilizer`. Let's only call it once, and after it's called, put the data into `$data['_attachments']` if it's of the `file` or `image` type.

dickolsson commented on src/Normalizer/FileItemNormalizer.php:
```diff
@@ -0,0 +1,65 @@
+      // @todo Add 'revpos' value to the result array.
+      $result = array(
+        $key => array(
+          'content_type' => $file->getMimeType(),
```
<del>We should use the service `file.mime_type.guesser` (`Drupal\Core\File\MimeType\MimeTypeGuesser`) to map MIME type to a valid `Content-Type` header. Because MIME type and Content-Type are not exactly the same thing.</del>


Actually, I'm wrong. I think it's ok to use the MIME type as you do here.

jeqq committed:
Call field normalize method once.

dickolsson commented on src/Plugin/rest/resource/AttachmentResource.php:
```diff
@@ -0,0 +1,179 @@
+      throw new NotFoundHttpException();
+    }
+
+    $file = \Drupal::entityManager()->loadEntityByUuid('file', $file_uuid);
```
We should create a Param Converter for the `$file_uuid` so that we don't need to do `entity_load()` here. I split the original Entity UUID converter into two specific ones. But you can recover it from here: https://github.com/dickolsson/drupal-relaxed/blob/7ad657810ddb3f2c48bc7b66c9fbb730d84eddc1/src/ParamConverter/EntityUuidConverter.php

The new `EntityUuidConverter` can probably be made a bit simpler than the one that I pointed to in the link.

dickolsson replied:
Same here, we should use a Param Converter for the `$file_uuid` parameter, so that it automatically becomes `$file`.

dickolsson replied:
Actually, the new `EntityUuidConverter` can almost be as simple as the new `DbConverter` (only use `->loadByUuid()` instead of `->load()`): https://github.com/dickolsson/drupal-relaxed/blob/8.x-1.x/src/ParamConverter/DbConverter.php

dickolsson replied:
And to make it such that each endpoint resource doesn't need to define the parameter type (like we had to do before, [see this](https://github.com/dickolsson/drupal-relaxed/blob/7ad657810ddb3f2c48bc7b66c9fbb730d84eddc1/src/Plugin/rest/resource/DocResource.php#L23)) we need to make so that we name the parameter `{entity_uuid:file}` instead of `{uuid}` in the route. Because then we can add the parameter type dynamically to the route. See how I do it for the `DocIdConverter` and `DbConverter` [here](https://github.com/dickolsson/drupal-relaxed/blob/8.x-1.x/src/Plugin/rest/resource/ResourceBase.php#L44).

dickolsson replied:
We can talk more about this on Monday, since it's a bit complicated :-)

jeqq committed:
Rename the service for FileItemNormalizer.

jeqq committed:
Add Param Converter for file uuid, use file.mime_type.guesser service.

dickolsson commented on src/Plugin/rest/resource/ResourceBase.php:
```diff
@@ -41,7 +41,7 @@ public function routes() {
       // @todo Consider converting this into something more generic.
       $parameters = array();
-      foreach (array('db', 'docid') as $parameter) {
+      foreach (array('db', 'docid', 'attachment') as $parameter) {
```
When we create a generic `EntityUuidConverter` this should become `{entity_uuid:file}`. This will indicate that the converter is of the primary type "entity_uuid" and will give the converter information about the entity type as well ("file" in this case).

jeqq committed:
Use getMimeType() instead of mime type guesser.

dickolsson commented:
Generally for the attachment endpoint we need to look into how to allow more content/mime type formats. The whole API only supports JSON right now. We might have to write some additional plugin or something... I'm not sure.

dickolsson commented:
I looked into the format thing a bit. What we need to do is to create a new attachment normalizer that only supports the `File` entity type class. This normalizer should have all possible file types as supported serialization format. This was the format will be allowed by the route. See `Drupal\relaxed\Plugin\rest\resource\ResourceBase` where the actual URL route is generated for all endpoints and how its using the serialized formats to set the content type requirements.

dickolsson commented:
So rather than doing the base64 encoding directly in the resource class it should be done in the serializer. Exactly as the Doc and Db resource work.

Commit:
Made the EntityUuidConverter work and implemented temporary route format and encoding support.

Commit:
Implemented AttachmentNormalizer and StreamEncoder to deal with attachment functionality

Commit:
Initial working version of the GET and HEAD methods for the attachment resource

Commit:
Fixed the PUT method for the attachment resource. Plus some late night refactoring to make the resources leaner.

Commit:
Fixed wrong variable name in the local doc resource

dickolsson commented:
Now it's only the `DELETE` method that remains to sort out...

Commit:
Fixed the DELETE method for the attachment resource

Commit:
Assert content-type header for all HEAD requests

Commit:
# This is a combination of 3 commits.
# The first commit's message is:
Assert content-type header for all HEAD requests

# This is the 2nd commit message:

Corrected the default response format

# This is the 3rd commit message:

Revert "Corrected the default response format"

This reverts commit 7eb323f9ec5b6794fbd4adfc5a768352e6069103.

dickolsson commented:
I think this commit will come back green. I will merge this if it does...
